[api-minor] Get rid of CSS transform on each annotation in the annotation layer

calixteman opened the pull request:
- each annotation has its coordinates/dimensions expressed in percentage,
  hence it's correctly positioned whatever the scale factor is;
- the font sizes are expressed in percentage too and the main font size
  is scaled thanks a css var (--scale-factor);
- the rotation is now applied on the div annotationLayer;
- this patch improve the rendering of some strings where the glyph spacing
  was not correct (it's a Firefox bug);
- it helps to simplify the code and it should slightly improve the update of
  page (on zoom or rotation).

calixteman committed:
Get rid of CSS transform on each annotation in the annotation layer

- each annotation has its coordinates/dimensions expressed in percentage,
  hence it's correctly positioned whatever the scale factor is;
- the font sizes are expressed in percentage too and the main font size
  is scaled thanks a css var (--scale-factor);
- the rotation is now applied on the div annotationLayer;
- this patch improve the rendering of some strings where the glyph spacing
  was not correct (it's a Firefox bug);
- it helps to simplify the code and it should slightly improve the update of
  page (on zoom or rotation).

Snuffleupagus commented:
I'm assuming, based on a very brief look at the code, that we no longer can easily avoid having both a `--scale-factor` and a `--zoom-factor` CSS variable?

Snuffleupagus commented on src/display/annotation_layer.js:
```diff
@@ -210,27 +214,6 @@ class AnnotationElement {
       page.view[3] - data.rect[3] + page.view[1],
     ]);
 
-    if (data.hasOwnCanvas) {
```
I could be wrong, but it seems that the `data.hasOwnCanvas`-property is effectively unused now.

Hence I wonder if we could perhaps use e.g. a `this._hasOwnCanvas`-property on the Annotation-instances, in `src/core/annotation.js`, to hold the value instead?

calixteman replied:
It's still used in `canvas.js` in order to know that the annotation must be rendered on its own canvas!

Snuffleupagus replied:
Yes, but that value [comes from this code](https://github.com/mozilla/pdf.js/blob/89cebcb6f9b757583bc855bf0ddbde857ed6308b/src/core/annotation.js#L872-L878), and it seems to me that you could perhaps [change this code](https://github.com/mozilla/pdf.js/blob/89cebcb6f9b757583bc855bf0ddbde857ed6308b/src/core/annotation.js#L851-L852) to:
```javascript
const isUsingOwnCanvas = 
   this._hasOwnCanvas && intent & RenderingIntentFlag.DISPLAY; 
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -129,6 +129,27 @@ class AnnotationLayerBuilder {
+      style.height = `${width}px`;
+    }
+
+    const { classList } = this.div;
+    for (const angle of [90, 180, 270]) {
+      const rotate = `rotate${angle}`;
+      if (angle === rotation) {
+        classList.add(rotate);
+      } else {
+        classList.remove(rotate);
+      }
+    }
```
*Possibly a stupid idea, but:* Could we perhaps simplify this a little bit, by using an attribute instead here?
```javascript
this.div.setAttribute("data-annot-rotation", rotation);
```
And the CSS file would then look something like:
```CSS
[data-annot-rotation="90"] {
  transform: rotate(90deg) translateY(-100%);
}
[data-annot-rotation="180"] {
  transform: rotate(180deg) translate(-100%, -100%);
}
[data-annot-rotation="270"] {
  transform: rotate(270deg) translateX(-100%);
}
```

Please note that the above is untested, but I wanted to ask nonetheless.

calixteman replied:
Good idea and it works :)

Snuffleupagus commented:
Also, with the last round of changes it seems that https://github.com/mozilla/pdf.js/blob/89cebcb6f9b757583bc855bf0ddbde857ed6308b/src/display/editor/annotation_editor_layer.js#L426-L432 is now unused. Are you planning on using that one later, or can it be removed?

Snuffleupagus commented:
/botio test

calixteman commented:
/botio browsertest

Snuffleupagus commented on src/display/base_factory.js:
```diff
@@ -143,14 +143,18 @@ class BaseSVGFactory {
   }
 
-  create(width, height) {
+  create(width, height, noDims = true) {
```
This changes the default behaviour of these classes, which could very well cause issues elsewhere. Furthermore, having an optional parameter where `undefined` implies `true` feels slightly unfortunate.

With that in mind, can we please change this to e.g. `create(width, height, skipDims = false) {` instead (with suitable changes below)?

calixteman replied:
My bad, I thought `false` and I wrote `true`.

calixteman commented:
/botio browsertest

Snuffleupagus commented on src/display/annotation_layer.js:
```diff
@@ -210,41 +214,13 @@ class AnnotationElement {
       const verticalRadius = data.borderStyle.verticalCornerRadius;
       if (horizontalRadius > 0 || verticalRadius > 0) {
-        const radius = `${horizontalRadius}px / ${verticalRadius}px`;
+        const radius = `calc(${horizontalRadius}px * var(--scale-factor)) / calc(${verticalRadius}px * var(--scale-factor))`;
```
Why do we need to multiply with `var(--scale-factor)` in both the numerator *and* denominator, since I'd expect that to cancel out?

calixteman replied:
I don't think the slash is used for a division:
https://developer.mozilla.org/en-US/docs/Web/CSS/border-radius#syntax
but  it's only a way to define two radius in order to have an elliptic curvature.
But maybe I overlooked something...

Snuffleupagus replied:
Sorry, I got lost in the parenthesis (since Prettier doesn't required a line-break here).

Snuffleupagus commented on src/display/annotation_layer.js:
```diff
@@ -1942,7 +1907,7 @@ class LineAnnotationElement extends AnnotationElement {
     const data = this.data;
     const { width, height } = getRectDims(data.rect);
-    const svg = this.svgFactory.create(width, height);
+    const svg = this.svgFactory.create(width, height, /* noDims = */ true);
```
Here, and elsewhere, please update all of these to mention `skipDims` in the comment instead.

Snuffleupagus commented:
Unfortunately there still appear to be a fair number of regressions here, based on the ref-tests:
 - Popup sizes, since they're *all* smaller now.
 - Popup *vertical* positions, since many of the are moved; e.g. `annotation-line-page1`.
 - Border widths becoming thinner; e.g. `annotation-border-styles.pdf-page1` and `bug1219400-page1`.
 - Annotations being rendering in the wrong spot; e.g. `issue4872-page1`.
 - Text sizes being smaller in forms; e.g. `annotation-choice-widget-forms-page1`.

calixteman commented:
* about the popup width: I just fixed few issues with popup width, but they'll be slightly different because the glyph spacing is not exactly the same: this is something this patch aims to fix so I think it's acceptable.
 * about the border widths: I compared output with Acrobat and I've the feeling it's an improvement. For example in `link-annotation-border.pdf`, on page 26, here is what I get with nightly:
  
![image](https://user-images.githubusercontent.com/5641725/173625215-6f610eb2-c80f-404c-a8b3-b1581d5728fe.png)
and now with the patch
 
![image](https://user-images.githubusercontent.com/5641725/173625497-3c9ac54d-9c9f-4495-82f1-bc844eccc7ae.png)

which very close to what Acrobat is rendering.
 * about `issue4872.pdf`: it's an error in test stuff: I fixed `annotation_layer_builder_overrides.css`
 * about popup vertical positions: for example
   - in `InkAnnotationElement`: the popup is appended to the container before the svg, hence the popup is on top-left of the annotation
   - in `LineAnnotationElement`: it's the opposite, hence the popup is on the bottom-left.

   So my feeling is that the vertical position is just a bit random. With this patch it'll be always on the top-left which is maybe not ideal but at least it's consistent.
   I think we should position the popup in the quarter we've the most of available area... but we can discuss about that in an issue.

calixteman commented:
/botio browsertest

calixteman commented:
There is always a font size issue in popups but it's likely on the test's side: I'll fix.

calixteman commented:
/botio browsertest

Snuffleupagus commented on src/display/annotation_layer.js:
```diff
@@ -210,41 +214,13 @@ class AnnotationElement {
-
     if (!ignoreBorder && data.borderStyle.width > 0) {
-      container.style.borderWidth = `${data.borderStyle.width}px`;
-      if (data.borderStyle.style !== AnnotationBorderStyleType.UNDERLINE) {
-        // Underline styles only have a bottom border, so we do not need
-        // to adjust for all borders. This yields a similar result as
-        // Adobe Acrobat/Reader.
-        width -= 2 * data.borderStyle.width;
-        height -= 2 * data.borderStyle.width;
-      }
```
What "changed" in the new implementation to make this hack unnecessary?
Also, do we have proper test-coverage for this code-path?

calixteman replied:
I added:
https://github.com/mozilla/pdf.js/pull/15036/files#diff-4748c83592581d85910b4a948677ddfcd25c8dc68ffcd20c524e0f4f0739f51aR53

and I compared the output of `annotation-border-styles.pdf` (the code path is taken here except in the underline case) in FIrefox and in Acrobat: the rendering is almost the same (even for the underline case).

calixteman commented:
/botio browsertest

calixteman commented:
I had a look on all the "regressions" and my feeling is that overall it's an improvement.
I suppose that the position of a popup or its exact width doesn't really matter but it can be a matter of taste, I've no real opinion on that.

Snuffleupagus commented:
> I suppose that the position of a popup or its exact width doesn't really matter but it can be a matter of taste, I've no real opinion on that.

While the latest version of the patch looks much better (when looking at the reference tests) and smaller changes are generally no problem, there's one thing that I still wonder here:
Ignoring the positional changes of some Popups, why are a couple of them *considerably wider* than before?

It would be nice to understand more specifically what changed here, since it's not just font-related given that the line-breaks are in different positions now; for example `issue13915-page51`, `pr7352-page1`, `pr12564-page1`

marco-c commented:
It's very visible on `pr7352-page1` for example in the "annotation" string.
Before:
![immagine](https://user-images.githubusercontent.com/1616846/173868866-3213e5d6-4ad3-410b-ae4f-ce581ec41eeb.png)
After:
![immagine](https://user-images.githubusercontent.com/1616846/173868963-dce8e88e-5b20-4432-9043-646d75ba5857.png)

calixteman commented:
/botio browsertest

calixteman commented:
/botio-windows browsertest

calixteman commented:
I think we're almost good now :)

timvandermeij commented on src/display/base_factory.js:
```diff
@@ -143,14 +143,18 @@ class BaseSVGFactory {
   }
 
-  create(width, height) {
+  create(width, height, skipDims = false) {
```
I'd prefer we call this `skipDimensions` for better readability here and in the callsites.

timvandermeij reviewed (commented):
Fixes #14935 I presume? Looks good to me with these comments addressed.

calixteman commented:
/botio test

calixteman commented:
/botio makeref
